packages/ubuntu_desktop_installer/lib/pages/connect_to_internet/wifi_model.dart

```diff
@@ -0,0 +1,314 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:collection/collection.dart';
+import 'package:dbus/dbus.dart';
+import 'package:flutter/material.dart';
+
+import '../../services.dart';
+import 'connect_model.dart';
+import 'network_device.dart';
+import 'property_stream_notifier.dart';
+
+// TODO: appropriate timeout?
+@visibleForTesting
+const kWifiScanTimeout = Duration(milliseconds: 7500);
+
+class WifiModel extends PropertyStreamNotifier implements ConnectModel {
+  WifiModel(this._service, [this._udev]) {
+    addProperties(_service.propertiesChanged);
+    addPropertyListener('Devices', _updateDevices);
+    addPropertyListener('WirelessEnabled', notifyListeners);
+  }
+
+  @override
+  bool get canConnect => _selected?.canConnect == false;
+
+  @override
+  bool get canContinue => _selected?.canConnect == true;
+
+  @override
+  bool get isActive => devices.any((device) => device.isActive);
+
+  @override
+  bool get isBusy => _selected?.isBusy == true || _selected?.scanning == true;
+
+  @override
+  bool get isEnabled => _service.wirelessEnabled;
+
+  @override
+  ConnectMode get connectMode => ConnectMode.wifi;
+
+  @override
+  void onSelected() {
+    final device = devices.firstWhereOrNull((device) {
+      return device.activeAccessPoint != null;
+    });
+    device?.init();
+    selectDevice(device);
+  }
+
+  @override
+  void onDeselected() {
+    if (_selected?.isBusy != true) return;
+    _selected!.disconnect();
+  }
+
+  @override
+  Future<void> init() async => _updateDevices();
+
+  @override
+  void dispose() {
+    _resetDevices();
+    super.dispose();
+  }
+
+  @override
+  Future<void> enable() {
+    log.debug('Enable wireless networking');
+    return _service.setWirelessEnabled(true);
+  }
+
+  @override
+  Future<void> connect() {
+    final device = selectedDevice!;
+    final accessPoint = device.selectedAccessPoint!;
+    log.debug('Connect $device to $accessPoint');
+    return _service.addAndActivateConnection(
+      connection: <String, Map<String, DBusValue>>{},
+      device: device.device,
+      accessPoint: accessPoint.accessPoint,
+    );
+  }
+
+  final NetworkService _service;
+  final UdevService? _udev;
+
+  List<WifiDevice>? _devices;
+  List<WifiDevice> get devices => _devices ??= _getDevices();
+
+  List<WifiDevice> _getDevices() {
+    final devices = _service.wirelessDevices
+        .map((device) => WifiDevice(device, _udev))
+        .toList();
+    log.debug(() => 'Update Wi-Fi devices: ${describeDevices(devices)}');
+    return devices;
+  }
+
+  void _resetDevices() {
+    if (_devices == null) return;
+    for (final device in _devices!) {
+      device.dispose();
+    }
+    _devices = null;
+  }
+
+  void _updateDevices() {
+    _resetDevices();
+    notifyListeners();
+  }
+
+  WifiDevice? _selected;
```

Comment:
There are a bunch of references to `_selected` in getters and methods which don't change its value, for consistency and improved readability they should probably be replaced by `selectedDevice`.